### PR TITLE
fix pipe read limit

### PR DESCRIPTION
**Pull request by jcupitt**
- We were not checking `VIPS_PIPE_READ_LIMIT` for huge single reads
- Use vips__parse_size() to parse the size of the read limit, so eg. "2g" now works for a 2gb read limit
- vipsheader was not always displaying the error buffer correctly

see https://github.com/libvips/libvips/issues/3984

### CODE REVIEW

**Comment by jcupitt** [comment]
Values of `VIPS_PIPE_READ_LIMIT` over 4gb will cause strange errors. With a 12gb test file:

```
$ export VIPS_PIPE_READ_LIMIT=4g
$ cat Visium_HD_Human_Colon_Cancer_tissue_image.btf | vipsheader -a stdin

(vipsheader:169956): GLib-ERROR **: 19:13:03.366: adding 4096 to array would overflow
Trace/breakpoint trap (core dumped)
```

Since GByteArray is limited to 4gb. Perhaps we should enforce 2**32-1 as an upper bound for `VIPS_PIPE_READ_LIMIT`?

**Comment by jcupitt** [comment]
... in case anyone wants to test this, you can gdb with a piped input with:

```
$ mkfifo my-fifo
$ cat hugefile > my-fifo &
$ gdb vipsheader
(gdb) run < my-fifo
```

(took me a while to work out)

File: ChangeLog [hunk 1]
```diff
@@ -20,6 +20,7 @@ master 8.16
 - webpsave has @passes parameter to set number of passes to achieve
   desired target_size [john-parton]
 - revise sRGB -> B_W coefficients [daniellovera]
+- fix pipe read limit
```
Comment (lovell) [inline ChangeLog:23]: Is it worth mentioning that the previous default limit of 1GB has been removed (and must now be set explicitly if required)?

Comment (kleisauke) [reply ChangeLog:23]: It looks like the default is still 1GiB, unless I'm overlooking something:
https://github.com/libvips/libvips/blob/9de5f24dab4ea2064666b351960a9719f27a4ac5/libvips/iofuncs/source.c#L79-L84

Comment (lovell) [reply ChangeLog:23]: D'oh, yes it is, thank you Kleis.

**Review by lovell — APPROVED** [review]

**Comment by kleisauke** [comment]
Should this PR target the [`8.15`](https://github.com/libvips/libvips/tree/8.15) branch instead? FWIW, version 8.15.3 has not been released yet; the release date in the changelog could be a bit misleading.
https://github.com/libvips/libvips/blob/a4be03b592204a95dec1b6eae4bf29e5efdabf2a/ChangeLog#L1

**Comment by jcupitt** [comment]
Ah true. I was imagining releasing 8.16 soon (it's been six months), but you're right that 8.15 has a couple of useful things that should go out before then. I'll retarget it.

**Comment by jcupitt** [comment]
Haha I completely failed to retarget it, I wonder how your are supposed to do this? Anyway, I've made a new pr for 8.15.

https://github.com/libvips/libvips/pull/3990

**Comment by kleisauke** [comment]
> I wonder how your are supposed to do this?

It's a bit hidden in GitHub's UI:
Change the base branch of a PR
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request

**Comment by jcupitt** [comment]
Ahh, I missed that. Thanks!

**Pull request closed without merging**